File: sEngSoftware/singleton.py
class Animal():
    def __init__(self, nome, idade, nomeTutor, porte):
        self.nome = nome
        self.idade = idade
        self.nomeTutor = nomeTutor
        self.porte = porte
    def comunicacao(self):
        print("Sou um animal, portanto, faço barulho.")
    def comportamento(self):
        print("Essa é minha forma de agir.")

class Cachorro(Animal):
    _instance = None
    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super(Cachorro, cls).__new__(cls)
        return cls._instance
    def __init__(self, nome, idade, porte, nomeTutor):
        if not hasattr(self, '_initialized'):
            super().__init__(nome, idade, nomeTutor, porte)
            self._initialized = True
    def comunicacao(self):
        print(f"Meu nome é {self.nome} e meu dono é {self.nomeTutor}.")
    def comportamento(self):
        print(f"Gosto de latir e de brincar com meu dono {self.nomeTutor}.")

class Gato(Animal):
    _instance = None
    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super(Cachorro, cls).__new__(cls)
        return cls._instance
    def __init__(self, nome, idade, porte, nomeTutor):
        if not hasattr(self, '_initialized'):
            super().__init__(nome, idade, nomeTutor, porte)
            self._initialized = True
    def comunicacao(self):
        print(f"Meu nome é {self.nome} e meu dono é {self.nomeTutor}.")
    def comportamento(self):
        print(f"Gosto de dormir e de ignorar meu dono {self.nomeTutor}.")

class Passaro(Animal):
    _instance = None
    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super(Cachorro, cls).__new__(cls)
        return cls._instance
    def __init__(self, nome, idade, porte, nomeTutor):
        if not hasattr(self, '_initialized'):
            super().__init__(nome, idade, nomeTutor, porte)
            self._initialized = True
    def comunicacao(self):
        print(f"Meu nome é {self.nome} e meu dono é {self.nomeTutor}.")
    def comportamento(self):
        print(f"Gosto muito de cantar para meu dono {self.nomeTutor}.")

c1 = Cachorro("Fred", 5, "Victor", "Médio")
c2 = Cachorro("Charles", 8, "Jaime", "Grande")
print(c1 is c2)
c1.comunicacao()
c2.comunicacao()
c1.comportamento()
c2.comportamento()

g1 = Gato("Cecília", 3, "Gabriela", "Pequeno")
g2 = Gato("Sophia", 14, "Natasha", "Médio")
print(g1 is g2)
g1.comunicacao()
g2.comunicacao()
g1.comportamento()
g2.comportamento()

p1 = Passaro("Zezé", 12, "Fabio", "Médio")
p2 = Passaro("Marcelo", 16, "Fernanda", "Pequeno")
print(p1 is p2)
p1.comunicacao()
p2.comunicacao()
p1.comportamento()
p2.comportamento()

File: baboseira/javascript/singleton.js
class Animal {
    constructor(nome, idade, nomeTutor, porte) {
        this.nome = nome;
        this.idade = idade;
        this.nomeTutor = nomeTutor;
        this.porte = porte;
    }
    comunicacao() {
        console.log("Sou um animal, portanto, faço barulho.");
    }
    comportamento() {
        console.log("Essa é minha forma de agir.");
    }
}

class Cachorro extends Animal {
    static instance = null;
    constructor(nome, idade, nomeTutor, porte) {
        if(!Cachorro.instance) {
            super(nome, idade, nomeTutor, porte);
            Cachorro.instance = this;
        }
        return Cachorro.instance;
    }
    comunicacao() {
        console.log(`Meu nome é ${this.nome} e meu dono é ${this.nomeTutor}.`);
    }
    comportamento() {
        console.log(`Gosto de latir e de brincar com meu dono ${this.nomeTutor}.`);
    }
}

class Gato extends Animal {
    static instance = null;
    constructor(nome, idade, nomeTutor, porte) {
        if(!Gato.instance) {
            super(nome, idade, nomeTutor, porte);
            Gato.instance = this;
        }
        return Gato.instance;
    }
    comunicacao() {
        console.log(`Meu nome é ${this.nome} e meu dono é ${this.nomeTutor}.`);
    }
    comportamento() {
        console.log(`Gosto de dormir e de deixar meu dono ${this.nomeTutor} no vácuo.`);
    }
}

class Passaro extends Animal { 
    static instance = null;
    constructor(nome, idade, nomeTutor, porte) {
        if(!Passaro.instance) {
            super(nome, idade, nomeTutor, porte);
            Passaro.instance = this;
        }
        return Passaro.instance;
    }
    comunicacao() {
        console.log(`Meu nome é ${this.nome} e meu dono é ${this.nomeTutor}.`);
    }
    comportamento() {
        console.log(`Gosto muito de cantar para meu dono ${this.nomeTutor}.`);
    }
}

const c1 = new Cachorro("Leo", 8, "Francisco", "Grande");
const c2 = new Cachorro("Bibi", 5, "Roberto", "Pequeno");
console.log(c1 === c2);
c1.comunicacao();
c2.comunicacao();
c1.comportamento();
c2.comportamento();

const g1 = new Gato("Neve", 5, "William", "Pequeno");
const g2 = new Gato("Fuligem", 10, "Gustavo", "Médio");
console.log(g1 === g2);
g1.comunicacao();
g2.comunicacao();
g1.comportamento();
g2.comportamento();

const p1 = new Passaro("Marcelo", 13, "Gabriel", "Pequeno");
const p2 = new Passaro("Dida", 23, "Daniel", "Grande");
console.log(p1 === p2);
p1.comunicacao();
p2.comunicacao();
p1.comportamento();
p2.comportamento();